### DISCUSSION TITLE
Unable to build Furo docs - getting `exception: '_static/styles/furo.css' is not in list`

### DISCUSSION THREAD

**Discussion by datajoely** [pradyunsg/furo#299] [Potential Issue] [answered]
Hello I work on the [Kedro](https://kedro.readthedocs.io/) project and I'm trying to see if we could switch our docs to furo but I keep getting the error below - any ideas what I could do to fix it? 

```Traceback (most recent call last):
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/furo/__init__.py", line 154, in _html_page_context
    _add_asset_hashes(
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/furo/__init__.py", line 139, in _add_asset_hashes
    index = static.index("_static/" + asset)
ValueError: '_static/styles/furo.css' is not in list

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/sphinx/cmd/build.py", line 280, in build_main
    app.build(args.force_all, filenames)
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/sphinx/application.py", line 346, in build
    self.builder.build_all()
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 259, in build_all
    self.build(None, summary=__('all source files'), method='all')
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 360, in build
    self.write(docnames, list(updated_docnames), method)
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 534, in write
    self._write_serial(sorted(docnames))
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 544, in _write_serial
    self.write_doc(docname, doctree)
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/sphinx/builders/html/__init__.py", line 603, in write_doc
    self.handle_page(docname, ctx, event_arg=doctree)
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/sphinx/builders/html/__init__.py", line 1016, in handle_page
    newtmpl = self.app.emit_firstresult('html-page-context', pagename,
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/sphinx/application.py", line 467, in emit_firstresult
    return self.events.emit_firstresult(event, *args,
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/sphinx/events.py", line 127, in emit_firstresult
    for result in self.emit(name, *args, allowed_exceptions=allowed_exceptions):
  File "/Users/$user/miniconda3/envs/kedro-dev/lib/python3.8/site-packages/sphinx/events.py", line 117, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function _html_page_context at 0x7fdf0e95d700> for event 'html-page-context' threw an exception (exception: '_static/styles/furo.css' is not in list)

Extension error:
Handler <function _html_page_context at 0x7fdf0e95d700> for event 'html-page-context' threw an exception (exception: '_static/styles/furo.css' is not in list)
```

If I inspect `furo/__init__.py:136` at runtime the "furo.css" file doesn't exist in the list:
![image](https://user-images.githubusercontent.com/35801847/141338587-335cbde6-cf20-491e-a572-58329d829ad8.png)

**Comment by pradyunsg** [comment]
How did you install Furo?

**Reply by pradyunsg** [reply]
If you could provide a series of steps to reproduce this issue, that'd be great.

**Reply by datajoely** [reply]
Hi @pradyunsg

I did `pip install furo` with no version restriction on python 3.8, it looks like I have `furo==2021.10.9` this version installed.

The steps for doing so are as follows, it's a lot of work - so help would be appreciated, but I was more interested if anyone had seen this before?

1. Clone/Fork [kedro](https://github.com/quantumblacklabs/kedro)
2. Create new 3.8 environment
3. Install furo
3. `make install-test-requirements`
4. Edit `docs/conf.py` to disable things not supported by furo:
     - Change line 175 to `furo`
     - Comment out lines 184
     - Comment out lines 214-217
5. Run `make devserver` to build the docs

**Reply by pradyunsg** [reply]
That looks a bit more involved than I'd like. My guess is that it is a configuration option/hook in your file that's messing stuff up.

Can you try commenting out basically all of your configuration file and try adding things in (bisected will be easiest) to find out what is involved? sphinx-autobuild should make that fairly not-painful to do.

**Reply by datajoely** [reply]
Thank you - thought as much, I was mostly asking just in-case this was a common error with a simple solution.

**Reply by pradyunsg** [reply]
I haven't seen this before, no.

**Reply by bertiewooster** [reply]
Hi @datajoely @pradyunsg was there a resolution to this? I'm getting the same error trying to use furo for [SymPy](https://www.sympy.org/en/index.html). Here's the process:

1. Add furo to doc/requirements.txt
2. In the doc folder, build dockerfile using `docker build -f Dockerfile.htmldoc -t sympy_htmldoc .` where Dockerfile.htmldoc is:

```
FROM python:3.8.12-slim-buster

RUN apt-get update
RUN apt-get install -y make librsvg2-bin imagemagick graphviz git

COPY requirements.txt /tmp
RUN pip install --upgrade pip
RUN pip install -r /tmp/requirements.txt

CMD cd /sympy/doc; make html
```

3. Build documentation using `docker run --rm -v /absolute/path/to/SYMPY_ROOT:/sympy sympy_htmldoc`

Error is:

```
Extension error (furo):
Handler <function _html_page_context at 0x7f674f1f6a60> for event 'html-page-context' threw an exception (exception: '_static/styles/furo.css' is not in list)
make: *** [Makefile:60: html] Error 2
```

**Reply by pradyunsg** [reply]
Can you run sphinx-build with --verbose?

**Comment by pradyunsg** [comment, marked as answer]
Figured out why this is happening! Unset `html_style` that is set in your conf.py file.

**Reply by pradyunsg** [reply]
The problematic line in the sympy configuration: https://github.com/sympy/sympy/blob/b0b66d5511575eda9f0e3c7cc048b7bf63a85706/doc/src/conf.py#L133

**Reply by pradyunsg** [reply]
Released with a better error message.

**Reply by pradyunsg** [reply]
https://github.com/pradyunsg/furo/commit/576b68e227adcb4d2a9ab0a5a21201af3390ba4b

**Reply by datajoely** [reply]
Thank you!

**Comment by bertiewooster** [comment]
@pradyunsg That worked--thanks for isolating the problem! I have to say that the [furo version of the SymPy docs](http://whitegloveapps.com/sympy/furo/index.html) looks great already--haven't even investigated doing any other tweaking to our existing conf.py.

**Comment by doyled-it** [comment]
I'm still getting this error. When I went in to debug this, I checked the value for all the `c.filename`s seen in [this check](https://github.com/pradyunsg/furo/blob/main/src/furo/__init__.py#L178C14-L178C14):

```python
['_static/pygments.css', '_static/styles/furo.css?digest=369552022d0b975c8e74270ce6eabe0fb7978f24', '_static/graphviz.css', '_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e']
```

It looks like you're appending an asset hash and it's missing it when it goes to check that list?

**Reply by pradyunsg** [reply]
Please open a new discussion, with a reproducer for the behaviour you're seeing. :)

**Reply by doyled-it** [reply]
Thanks for responding. I opened one [here](https://github.com/pradyunsg/furo/discussions/708).

**Comment by arxyzan** [comment]
@doyled-it Do you also have `sphinx_rtd_theme` installed along with `furo`? I had this problem when trying to deploy my docs on github pages and after deleting `sphinx_rtd_theme` from the requirements, the problem was solved.

**Reply by mcflugen** [reply]
@arxyzan Thank you! This was exactly my problem. You saved me so much time!

**Reply by doyled-it** [reply]
Hmm, that didn't end up working for me. I'm only experiencing this issue after doing a `poetry update` command last week.
